domain/model/service/service.go

```diff
@@ -709,3 +730,60 @@ func (s WatchableService) WatchModel(ctx context.Context, modelUUID coremodel.UU
 		}),
 	)
 }
+
+// WatchModelCloudCredential returns a new NotifyWatcher watching for changes that
+// result in the cloud spec for a model changing. The changes watched for are:
+// - updates to model cloud.
+// - updates to model credential.
+// - changes to the credential set on a model.
+// The following errors can be expected:
+// - [modelerrors.NotFound] when the model is not found.
+func (s *WatchableService) WatchModelCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (watcher.NotifyWatcher, error) {
+	if err := modelUUID.Validate(); err != nil {
+		return nil, errors.Errorf("invalid model UUID watching model cloud credential: %w", err)
+	}
+
+	// Get the model's cloud and credential UUID to use in the watcher filters.
+	cloudUUID, credentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+	if err != nil {
+		return nil, errors.Errorf("getting model cloud and credential: %w", err)
+	}
+
+	var currentCredentialUUID atomic.Pointer[credential.UUID]
+	currentCredentialUUID.Store(&credentialUUID)
+
+	// The mapper is used to filter out any model events where the credential UUID has not changed.
+	mapper := func(ctx context.Context, txnRunner database.TxnRunner, events []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		// Get the current model credential UUID.
+		_, newCredentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		gotNewCredential := *currentCredentialUUID.Load() != newCredentialUUID
+		var out []changestream.ChangeEvent
+		for _, e := range events {
+			if e.Namespace() != "model" || gotNewCredential {
+				out = append(out, e)
+			}
+		}
+		currentCredentialUUID.Store(&newCredentialUUID)
+		return out, nil
+	}
+
+	result, err := s.watcherFactory.NewNotifyMapperWatcher(
+		mapper,
+		eventsource.PredicateFilter("model", changestream.Changed, func(s string) bool {
+			return s == modelUUID.String()
+		}),
+		eventsource.PredicateFilter("cloud", changestream.Changed, func(s string) bool {
+			return s == cloudUUID.String()
+		}),
+		eventsource.PredicateFilter("cloud_credential", changestream.Changed, func(s string) bool {
+			return s == currentCredentialUUID.Load().String()
+		}),
+	)
+	if err != nil {
+		return result, errors.Errorf("watching model cloud and credential: %w", err)
```

Comment:
Nit: Do you think if it's better to return nil if there's any err regardless of how the downstream function handles the return value for any err?

Reply:
yup